stock_putaway_recursive/tests/test_stock_putaway_recursive.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo.tests import SavepointCase
+
+
+class TestStockPutawayRecursive(SavepointCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
+        cls.product_flipover = cls.env.ref('product.product_product_20')
+        cls.order_location = cls.env.ref('stock.location_order')
+        cls.stock_location = cls.env.ref('stock.stock_location_stock')
+        cls.shelf1_location = cls.env.ref('stock.stock_location_components')
+        cls.shelf2_location = cls.env.ref('stock.stock_location_14')
+        cls.refri_location = cls.env.ref('stock.location_refrigerator_small')
+
+        # Define putaway from Stock to Stock/Shelf 1
+        stock_putaway = cls.env['product.putaway'].create({
+            'name': 'stock putaway',
+        })
+        cls.stock_location.write({
+            'putaway_strategy_id': stock_putaway.id
+        })
+        stock_putaway.write({
+            'product_location_ids': [(0, 0, {
+                'product_id': cls.product_flipover.id,
+                'fixed_location_id': cls.shelf1_location.id,
+            })]
+        })
+        # Define putaway from Stock/Shelf 1 to Stock/Shelf 2
+        shelf1_putaway = cls.env['product.putaway'].create({
+            'name': 'shelf1 putaway',
+        })
+        cls.shelf1_location.write({
+            'putaway_strategy_id': shelf1_putaway.id
+        })
+        shelf1_putaway.write({
+            'product_location_ids': [(0, 0, {
+                'product_id': cls.product_flipover.id,
+                'fixed_location_id': cls.shelf2_location.id,
```

Comment:
Domain in https://github.com/odoo/odoo/blob/12.0/addons/stock/views/product_strategy_views.xml#L18 should prohibit this.